### ISSUE TITLE
add rule: new() in an interface must not return the same interface.

### ISSUE THREAD

**Issue by JeroMiya** [palantir/tslint#261]
One common pattern I see in TypeScript type declarations is the inclusion of a new(args...): IInterface in an interface to type a constructor and the thing it constructs in the same interface:

``` typescript
interface IConstructable {
  new(): IConstructable;
  field: string;
}
declare var Constructable: IConstructable;
var x = new Constructable();
var y = new x(); // oops! I didn't intend this to be possible!
```

As you can see, this leads to a recursive type. Probably not what you wanted. This rule would enforce that new() in an interface must not return that interface. The right pattern to follow is to separate out the new() into another interface:

``` typescript
interface IStringValueBox {
  value: string;
}

interface IStringValueBoxConstructor {
  new(): IStringValueBox;
}

declare var StringValueBox: IStringValueBoxConstructor;
var x = new StringValueBox();
var y = new x(); // Error, IStringValueBox does not have a constructor, etc...
```

**Comment by aleksey-bykov** [comment]
What is your problem with constructors that by definition return the
interface they belong to or with recursive types for the same matter?

**Comment by JeroMiya** [comment]
Think of how you would have to actually implement this interface in JavaScript (remember you can't implement new() from an interface in a typescript class that 'implements' the interface explicitly):

```typescript
interface IConstructable {
    new(): IConstructable;
    field: string;
}

declare var Constructable: IConstructable;
alert(Constructable.field); // this passes the type checker
alert(new (new Constructable())); // this also passes the type checker
```

The only thing that could possibly implement this as defined would be something like this, and this code makes no sense:
```typescript
function Constructable() {
    return Constructable;
}
Constructable.field = "foo";
```
I've never seen a javascript constructor return a function, let alone return itself as a function, and I'm not sure what use case would necessitate that. Really, a constructor should never return anything - it should modify the passed in 'this' pointer only - otherwise you should be using a regular function not a constructor. If there was a reason for it, I'm sure they could just add a local tslint exception.

I don't have a problem in general with recursive types. It is just very unlikely for a constructor to be recursive in this way.

**Comment by awerlang** [comment]
This would clearly make no sense. As interfaces stands for the instance side of a class, it should have a new () method on it. And if it has, it should represent a factory, another object.

Could you point where it is becoming common?

**Comment by andy-hanson** [comment]
Should be fixed by #1963.
